activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/rest/TaskPermissionsHelper.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2017-2020 Alfresco Software, Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.activiti.cloud.services.query.rest;
+
+import org.activiti.api.runtime.shared.security.SecurityManager;
+import org.activiti.cloud.api.task.model.QueryCloudTask.TaskPermissions;
+import org.activiti.cloud.services.query.model.QTaskEntity;
+import org.activiti.cloud.services.query.model.TaskEntity;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TaskPermissionsHelper {
+    
+    private final SecurityManager securityManager;
+
+    private final TaskControllerHelper taskControllerHelper;
+
+    public TaskPermissionsHelper(SecurityManager securityManager, TaskControllerHelper taskControllerHelper) {
+        this.securityManager = securityManager;
+        this.taskControllerHelper = taskControllerHelper;
+    }
+
+    public void setCurrentUserTaskPermissions(TaskEntity task) {
+        String userId = securityManager.getAuthenticatedUserId();
+        if(userId!=null) {
+            List<String> userGroups = securityManager.getAuthenticatedUserGroups();
+            List<TaskPermissions> permissions = new ArrayList<>();
+            if(!canUserViewTask(task)) return;
+            
+            permissions.add(TaskPermissions.VIEW);
+            
+            if(canUserClaimTask(task, userId, userGroups)) {
+                permissions.add(TaskPermissions.CLAIM);
+            }
+            
+            if(isUserAssignee(task, userId)) {
+                permissions.add(TaskPermissions.RELEASE);
+                permissions.add(TaskPermissions.UPDATE);
+            }
+
+            task.setPermissions(permissions);
+        }
+    }
+    
+    private boolean canUserViewTask(TaskEntity task) {
+        return taskControllerHelper.canUserViewTask(QTaskEntity.taskEntity.id.eq(task.getId()));
+    }
+    
+    private boolean canUserClaimTask(TaskEntity task, String userId, List<String> userGroups) {
+        boolean isUserOwner = task.getOwner() != null && task.getOwner().equals(userId);
+        
+        return !isTaskAssigned(task) && (isUserOwner || isUserCandidate(task, userId, userGroups));
```

Comment:
task owner can't  claim a task only candidate users or members of candidate groups can claim a task

Reply:
In my understanding, the owner is the creator of a standalone task, and he can claim a task also if he is not a candidate user or member of a candidate group.

Reply:
Standalone tasks aren't claimable.

Reply:
Owner can reassign the task (that's how it works in APS)